File: apps/api/app/domain/embeddings.py
"""NVIDIA embeddings — notebook style: OpenAI client + get_embedding()."""

from __future__ import annotations

import asyncio

from openai import OpenAI

from app.core.config import get_settings

NVIDIA_DIM = 2048
DEFAULT_MODEL = "nvidia/nemotron-3-embed-1b"
DEFAULT_BASE = "https://integrate.api.nvidia.com/v1"

_openai: OpenAI | None = None


def embedding_configured() -> bool:
    settings = get_settings()
    return bool((settings.nvidia_api_key or settings.embedding_api_key or "").strip())


def embedding_dim() -> int:
    settings = get_settings()
    if settings.embedding_dim and settings.embedding_dim > 0:
        return settings.embedding_dim
    return NVIDIA_DIM if embedding_configured() else 256


def _openai_client() -> OpenAI:
    """Notebook: OpenAI(base_url=NVIDIA Integrate, api_key=NVIDIA_API_KEY)."""
    global _openai
    if _openai is not None:
        return _openai
    settings = get_settings()
    api_key = (settings.nvidia_api_key or settings.embedding_api_key or "").strip()
    if not api_key:
        raise RuntimeError("NVIDIA_API_KEY / EMBEDDING_API_KEY required for semantic embeddings")
    base = (settings.embedding_base_url or DEFAULT_BASE).rstrip("/")
    _openai = OpenAI(api_key=api_key, base_url=base, timeout=60.0)
    return _openai


def get_embedding(text: str, model: str | None = None) -> list[float]:
    """Notebook cell: openai.embeddings.create(input=text, model=...).data[0].embedding"""
    trimmed = (text or "").strip()
    dim = embedding_dim()
    if not trimmed:
        return [0.0] * dim
    settings = get_settings()
    use_model = model or settings.embedding_model or DEFAULT_MODEL
    response = _openai_client().embeddings.create(
        input=trimmed[:8000],
        model=use_model,
    )
    return list(response.data[0].embedding)


async def embed_query(text: str) -> list[float]:
    return await asyncio.to_thread(get_embedding, text)

File: apps/web/src/server/embeddings.ts
/**
 * Embeddings for semantic cache.
 * Prefer notebook model: nvidia/nemotron-3-embed-1b via NVIDIA Integrate API (2048-d).
 * Falls back to local hashed vectors (256-d) when no embedding API key is set.
 */

const NVIDIA_BASE = "https://integrate.api.nvidia.com/v1";
const DEFAULT_MODEL = "nvidia/nemotron-3-embed-1b";
export const NVIDIA_EMBED_DIM = 2048;
export const LOCAL_EMBED_DIM = 256;

export type EmbeddingBackend = "nvidia" | "local";

export function embeddingApiKey() {
  return (
    process.env.NVIDIA_API_KEY?.trim() ||
    process.env.EMBEDDING_API_KEY?.trim() ||
    ""
  );
}

export function embeddingBaseURL() {
  return process.env.EMBEDDING_BASE_URL?.trim().replace(/\/$/, "") || NVIDIA_BASE;
}

export function embeddingModel() {
  return process.env.EMBEDDING_MODEL?.trim() || DEFAULT_MODEL;
}

export function embeddingBackend(): EmbeddingBackend {
  return embeddingApiKey() ? "nvidia" : "local";
}

export function embeddingDim() {
  const forced = Number(process.env.EMBEDDING_DIM ?? "");
  if (Number.isFinite(forced) && forced > 0) return forced;
  return embeddingBackend() === "nvidia" ? NVIDIA_EMBED_DIM : LOCAL_EMBED_DIM;
}

export async function embedQuery(text: string, localEmbed: (t: string) => number[]): Promise<number[]> {
  const trimmed = text.trim();
  if (!trimmed) return new Array(embeddingDim()).fill(0);

  if (embeddingBackend() === "local") {
    return localEmbed(trimmed);
  }

  const response = await fetch(`${embeddingBaseURL()}/embeddings`, {
    method: "POST",
    headers: {
      "content-type": "application/json",
      authorization: `Bearer ${embeddingApiKey()}`,
    },
    body: JSON.stringify({
      input: trimmed.slice(0, 8000),
      model: embeddingModel(),
      encoding_format: "float",
    }),
  });

  if (!response.ok) {
    const detail = await response.text().catch(() => response.statusText);
    throw new Error(`Embedding failed (${response.status}): ${detail.slice(0, 300)}`);
  }

  const data = (await response.json()) as {
    data?: Array<{ embedding?: number[] }>;
  };
  const vector = data.data?.[0]?.embedding;
  if (!vector?.length) throw new Error("Embedding response missing vector");
  return vector;
}
